plugins/plugins.go

```diff
@@ -22,6 +22,30 @@ type Plugin interface {
 	Stop(ctx context.Context)
 }
 
+// PluginInitFunc is a function that will initialize a plugin given configuration document config
+// The function should return the plugin if successful and and error if unsuccessful
+type PluginInitFunc func(m *Manager, config []byte) (Plugin, error)
+
+var (
+	registeredPlugins    map[string]PluginInitFunc
+	registeredPluginsMux sync.Mutex
+)
+
+// RegisterPlugin registers a plugin with the plugins package. When a Runtime
+// is created, the factory functions will be called. This function is idempotent.
+func RegisterPlugin(name string, factory PluginInitFunc) {
+	registeredPluginsMux.Lock()
+	defer registeredPluginsMux.Unlock()
+	registeredPlugins[name] = factory
+}
+
+// GetRegisteredPlugins returns all the plugins that are currently registered with the plugins package.
```

Comment:
If the code is put back into the runtime package, I don't think this will be necessary. I'd rather remove it to reduce the # of APIs.

Reply:
See my comment above.